edb/server/protocol/binary.pyx

```diff
@@ -2048,10 +2049,23 @@ cdef class EdgeConnection:
 
         # number of elements in the tuple
         # for empty tuple it's okay to send zero-length arguments
-        # TODO(tailhook) force zero-length arguments for protocol 0.12 after
-        # upgrading the bindings
-        if frb_get_len(&in_buf) == 0:
-            recv_args = 0
+        if self.protocol_version >= (0, 12):
+            frb_len = frb_get_len(&in_buf)
+            is_null_type = \
+                query.query_unit.in_type_id == sertypes.NULL_TYPE_ID.bytes
+            if frb_len and is_null_type:
+                raise errors.ProtocolError(
+                    "absence of query arguments must be encoded with a 'zero' "
+                    "type (id: 00000000-0000-0000-0000-000000000000, "
+                    "encoded with zero bytes)")
+            elif not frb_len and not is_null_type:
+                raise errors.ProtocolError(
+                    f"insufficient data for type-id "
+                    f"{query.query_unit.in_type_id}")
+            elif frb_len == 0:
+                recv_args = 0
+            else:
+                recv_args = hton.unpack_int32(frb_read(&in_buf, 4))
```

Comment:
Please add a check that the input type id is a shape.

Reply:
(I just want this code path to be as easy to follow as possible, sorry for nit-picking)

Reply:
Is there a better way to do this than parsing `query.query_unit.in_type_data`?

Reply:
Oh, right, it's fine. Please fix the CI.

Reply:
> Please fix the CI.

Please approve this PR https://github.com/edgedb/edgedb-python/pull/240.